clients/drpoints/drpoints.cpp

```diff
@@ -0,0 +1,439 @@
+ * DAMAGE.
+ */
+
+/* DrPoints: Basic Block Vector (BBV) Client.
+ *
+ * Given a user-defined instruction interval, computes the BBVs (histogram of BB
+ * frequencies within the interval) of a program execution and outputs them in a .bb file.
+ */
+
+#include "dr_api.h"
+#include "dr_defines.h"
+#include "dr_events.h"
+#include "drmgr.h"
+#include "droption.h"
+#include "drreg.h"
+#include "drvector.h"
+#include "drx.h"
+#include "hashtable.h"
+
+namespace dynamorio {
+namespace drpoints {
+namespace {
+
+using ::dynamorio::droption::bytesize_t;
+using ::dynamorio::droption::DROPTION_SCOPE_CLIENT;
+using ::dynamorio::droption::droption_t;
+
+#define FATAL(...)                       \
+    do {                                 \
+        dr_fprintf(STDERR, __VA_ARGS__); \
+        dr_abort();                      \
+    } while (0)
+
+#define HASH_BITS 13
+
+#define MINSERT instrlist_meta_preinsert
+
+#ifdef X86_64
+#    define INLINE_COUNTER_UPDATE 1
+#else
+// TODO i#7685: We don't have the inlining implementation yet for AARCH64 or 32-bit
+// architectures.
+#endif
+
+static droption_t<bytesize_t> instr_interval(
+    DROPTION_SCOPE_CLIENT, "instr_interval", 100000000 /*=100M instructions*/,
+    "The instruction interval for which to generate BBVs",
+    "Divides the program execution in instruction intervals of the specified size and "
+    "generates BBVs as the BB execution frequency within the interval times the number "
+    "of instructions in the BB. Default is 100M instructions.");
+
+static droption_t<bool>
+    print_to_stdout(DROPTION_SCOPE_CLIENT, "print_to_stdout", false,
+                    "Enables printing the Basic Block Vectors to standard output",
+                    "Also prints to standard output on top of generating the .bbv file. "
+                    "Default is false.");
+
+static droption_t<bool> no_out_bbv_file(
+    DROPTION_SCOPE_CLIENT, "no_out_bbv_file", false,
+    "Disables the generation of the output .bbv file",
+    "Disables the generation of the output .bbv file, but still runs the client. Useful "
+    "for unit tests or paired with -print_to_stdout. Default is false.");
+
+// Global hash table that maps the PC of a BB's first instruction to a unique,
+// increasing ID that comes from unique_bb_count.
+static hashtable_t pc_to_id_map;
+
+// Global hash table to keep track of the execution count of BBs.
+// Key: unique BB ID, value: execution count in terms of BB instruction size.
+static hashtable_t bb_count_table;
+
+// Global unique BB counter used as ID. It must start from 1.
+static uint64_t unique_bb_count = 1;
+
+// Global instruction counter to keep track of when we reach the end of the user-defined
+// instruction interval.
+static uint64_t instr_count = 0;
+
+// List of Basic Block Vectors (BBVs).
+// This is a vector of vector pointers. Each element is a vector of pairs
+// <BB_ID, execution_count * BB_size> of type bb_id_count_pair_t.
```

Comment:
Explain that each element is for one instruction interval (right?).
Ordered in program order, right?

Reply:
Yes to both. Added to the comment.